_build/man/brainx.1
.\" Man page generated from reStructuredText.
.
.TH "BRAINX" "1" "June 22, 2014" "0.1" "Brainx"
.SH NAME
brainx \- Brainx Documentation
.
.nr rst2man-indent-level 0
.
.de1 rstReportMargin
\\$1 \\n[an-margin]
level \\n[rst2man-indent-level]
level margin: \\n[rst2man-indent\\n[rst2man-indent-level]]
-
\\n[rst2man-indent0]
\\n[rst2man-indent1]
\\n[rst2man-indent2]
..
.de1 INDENT
.\" .rstReportMargin pre:
. RS \\$1
. nr rst2man-indent\\n[rst2man-indent-level] \\n[an-margin]
. nr rst2man-indent-level +1
.\" .rstReportMargin post:
..
.de UNINDENT
. RE
.\" indent \\n[an-margin]
.\" old: \\n[rst2man-indent\\n[rst2man-indent-level]]
.nr rst2man-indent-level -1
.\" new: \\n[rst2man-indent\\n[rst2man-indent-level]]
.in \\n[rst2man-indent\\n[rst2man-indent-level]]u
..
.sp
This program can read and execute programs in BrainFuck, Brailoller and Braincopter languages.
.sp
Contents:
.SH BRAINX TUTORIAL
.sp
\fBusage\fP: brainx.py [\-h] [\-l | \-c | \-f] file|data
.sp
Execute Brain{fuck,loller,copter} code.
.INDENT 0.0
.TP
.B positional arguments:
file|data          input source file or actual source code
.TP
.B optional arguments:
.INDENT 7.0
.TP
.B \-h\fP,\fB  \-\-help
show this help message and exit
.TP
.B \-l\fP,\fB  \-\-brainloller
brainloller (default: false)
.TP
.B \-c\fP,\fB  \-\-braincopter
braincopter (default: false)
.TP
.B \-f\fP,\fB  \-\-brainfuck
brainfuck (default: true)
.UNINDENT
.UNINDENT
.sp
\fBExample files are located in test_data folder.\fP
.SH BRAINX  MEMBERS
.INDENT 0.0
.TP
.B class brainx.BrainCopter(filename)
BrainCopter interpreter.
.INDENT 7.0
.TP
.B operations(color)
Interprets BrainCopter code.
.UNINDENT
.UNINDENT
.INDENT 0.0
.TP
.B class brainx.BrainFuck(data, memory=b\(aqx00\(aq, memory_pointer=0)
Braifuck interpreter.
.INDENT 7.0
.TP
.B find_input()
Save chars after \(aq!\(aq and cut it off from source code.
.UNINDENT
.INDENT 7.0
.TP
.B get_loopcode(code)
Cut part of code to loop trough.
.UNINDENT
.INDENT 7.0
.TP
.B interpreter(code)
Define Brainfuck operations and execute them.
.UNINDENT
.INDENT 7.0
.TP
.B read_char()
Part of brainfuck extentsion, read 1 char either from input or from part after \(aq!\(aq.
.UNINDENT
.UNINDENT
.INDENT 0.0
.TP
.B class brainx.BrainLoller(filename)
BrainLoller interpreter.
.INDENT 7.0
.TP
.B decode(data)
Execute BrainLoller code.
.UNINDENT
.INDENT 7.0
.TP
.B move(x, y, d)
Change direction of instruction pointer.
.UNINDENT
.INDENT 7.0
.TP
.B operations(color)
Define operations by color of pixel.
.UNINDENT
.UNINDENT
.SH PNG PARSER  MEMBERS
.INDENT 0.0
.TP
.B exception image_png.PNGNotImplementedError
Exception informing that PNG image has structure which is not supported.
.UNINDENT
.INDENT 0.0
.TP
.B exception image_png.PNGWrongHeaderError
Exceptions informing that loaded file is probably not PNG image.
.UNINDENT
.INDENT 0.0
.TP
.B class image_png.PngReader(filepath)
Process PNG images.
.INDENT 7.0
.TP
.B bytes_to_num(bytes)
Convert 4 bytes to intiger value.
.UNINDENT
.INDENT 7.0
.TP
.B check_header()
Check header and cut it off
.UNINDENT
.INDENT 7.0
.TP
.B decode()
Save pixels according to used filetrs.
.UNINDENT
.INDENT 7.0
.TP
.B paeth(a, b, c)
Paeth predictor, part of 4th filter.
.UNINDENT
.INDENT 7.0
.TP
.B parse_png()
From raw binary data, get chunks and save them..
.UNINDENT
.UNINDENT
.INDENT 0.0
.IP \(bu 2
\fIgenindex\fP
.IP \(bu 2
\fImodindex\fP
.IP \(bu 2
\fIsearch\fP
.UNINDENT
.SH AUTHOR
Lubos Hercka
.SH COPYRIGHT
2014, Lubos Hercka
.\" Generated by docutils manpage writer.
.


brainx.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
import sys
import math
from image_png import PngReader
import argparse

class BrainFuck:
    """Braifuck interpreter."""
    
    def __init__(self, data, memory=b'\x00', memory_pointer=0):
        """Inicialization of Brainfuck interpreter."""
        
        # inicializace proměnných
        self.memory = bytearray(memory)
        self.memory_pointer = memory_pointer


        self.code = data
        
        # DEBUG a testy
        # a) paměť výstupu
        self.output = ''
        if self.code.count('[')!=self.code.count(']'):
            print ("Bad fromat")
            sys.exit(1)
        

        self.input=self.find_input()
        self.interpreter(self.code)
        

    def interpreter(self,code):
        """ Define Brainfuck operations and execute them."""
        code_ptr=0
        while code_ptr<len(code):
            command=code[code_ptr]
                
            if command==">":
                self.memory_pointer += 1

                if self.memory_pointer==len(self.memory):
                    self.memory+=bytearray([0])

            if command=="<":
                self.memory_pointer -= 1

                if self.memory_pointer<0:
                    self.memory_pointer=0

            if command=="+":
                self.memory[self.memory_pointer ] = (self.memory[self.memory_pointer ] + 1) % 256;

            if command=="-":
                if self.memory[self.memory_pointer]==0:
                    self.memory[self.memory_pointer]=255
                else:
                    self.memory[self.memory_pointer]-=1

            if command == ".":
                sys.stdout.write(chr(self.memory[self.memory_pointer]))
                towrite=chr(self.memory[self.memory_pointer])
                self.output+=towrite

            if command==",":
                self.memory[self.memory_pointer]=ord(self.read_char())

            if command=="[":
                loopcode = self.get_loopcode(code[code_ptr:])
                while self.memory[self.memory_pointer] != 0:
                    self.interpreter(loopcode)
                code_ptr += len(loopcode) + 1
            
            code_ptr += 1               

    def get_loopcode(self,code):
        """Cut part of code to loop trough."""    
        end = 1
        while (code[0:end].count('[') != code[0:end].count(']')):
            end += 1
        
       
        return code[1:end-1]
    def find_input(self):
        """Save chars after '!' and cut it off from source code."""
        if self.code.find('!')!=-1:
            pos=self.code.find('!')
            _input=self.code[pos+1:]
            self.code=self.code[:pos]

            return _input
        return 0

    def read_char(self):
        """Part of brainfuck extentsion, read 1 char either from input or from part after '!'."""
        if self.input!=0:
            ret=self.input[0]
            self.input=self.input[1:]
            return ret
        else:
            return sys.stdin.read(1)
        
    #
    # pro potřeby testů
    #
    def get_memory(self):
        # Nezapomeňte upravit získání návratové hodnoty podle vaší implementace!
        return self.memory


class BrainLoller():
    """BrainLoller interpreter."""
    
    def __init__(self, filename):
        """Inicialization of BrainLoller interpreter."""
        self.img=PngReader(filename).rgb
        # self.data obsahuje rozkódovaný zdrojový kód brainfucku..
        self.data = self.decode(self.img)
        # ..který pak předhodíme interpretru
        self.program = BrainFuck(self.data)

    
    def decode(self,data):
        """Execute BrainLoller code."""
        x=0
        y=0
        self.direction=2
        code=''
        while True:

            if x<0 or y<0 or x>=len(data) or y >=len (data[0]):
                break
            op=self.operations(data[x][y])
            if op[0]:
                code+=op[0]
            self.direction=op[1]
            x,y=self.move(x,y,self.direction)



        return code

    def operations (self,color):
        """Define operations by color of pixel."""
        command=''
        direction=self.direction
        if color == (255,0,0):
            command = '>'
        if color == (128,0,0):
            command = '<'
        if color == (0,255,0):
            command = '+'
        if color == (0,128,0):
            command = '-'
        if color == (0,0,255):
            command = '.'
        if color == (0,0,128):
            command = ','
        if color == (255,255,0):
            command = '['
        if color == (128,128,0):
            command = ']'
        if color == (0,255,255): # turn right
            direction=(direction+1)%4
        if color == (0,128,128): # turn left
            direction=(direction-1)%4

        return command,direction
    def move (self,x,y,d):
        """Change direction of instruction pointer."""
        if d==0:
            y-=1
        if d==1:
            x-=1
        if d==2:
            y+=1
        if d==3:
            x+=1
        return x,y

class BrainCopter(BrainLoller):
    """BrainCopter interpreter."""
    
    def operations(self,color):
        """Interprets BrainCopter code."""
        command = ''
        direction=self.direction
        color_c = (-2*color[0] + 3*color[1] + color[2])%11
        bf = '><+-.,[]'
        if color_c < 8:
            command = bf[color_c]
        if color_c == 8: # turn right
            direction=(direction+1)%4
        if color_c == 9: # turn left
            direction=(direction-1)%4
        return command,direction

def data_parser(data):
    try:
        with open(data, mode='r') as f:
            code = f.read()
            return code
    except:
        return data

if __name__ == '__main__':
    parser = argparse.ArgumentParser(description="Execute Brain{fuck,loller,copter} code.")
    group = parser.add_mutually_exclusive_group()
    parser.add_argument('data', metavar='file|data',
                     type=data_parser,
                     help='input source file or actual source code')
    
    group.add_argument('-l','--brainloller',
                     action='store_true',
                     dest='brainloller',
                     help='brainloller (default: false)')

    group.add_argument('-c','--braincopter',
                     action='store_true',
                     dest='braincopter',
                     help='braincopter (default: false)')
    group.add_argument('-f','--brainfuck',
                     action='store_true',
                     dest='brainfuck',
                     help='brainfuck (default: true)')
    args=parser.parse_args()
    if args.brainloller:
        BrainLoller(args.data)
    elif args.braincopter:
        BrainCopter(args.data)
    else:
        BrainFuck(args.data)
